aws/amazon-freertos
# CommonIO Flash: nrf52840-dk

**Description:**
Implementation of iot_flash.h

Async APIs are not supported in this PR. The memory mapped NVMC peripheral does not provide a mechanism for generating interrupts. NVMC is shared with CPU. CPU is halted if it's waiting to use NVMC. The _appearance_ of async can be done by launching a sync API in an RTOS thread.

Flash access protection is not supported in this PR, but I do have a working implementation in a separate branch. Excluded for now because testing is trickier and incomplete. For this board, flash protection violations cause hardfaults. There's no way to test via injected failure without overriding `HardFault_Handler` and restoring execution for test to catch failure. Need further discussion with team.

**Notes:**
- For this board, hardfaults are generated when certain flash violations or flash failures occur. These tests require a working flash page that is within life-span. 
 - Some of the CommonIO Flash tests erase/write pages thereby depleting flash life. Flash pages of this board are functional for a minimum of 10000 erase cycles.

**Testing:**
1) Set `testrunnerFULL_COMMON_IO_ENABLED = 1` in `aws_test_runner_config.h`
2) Set `IOT_TEST_COMMON_IO_FLASH_SUPPORTED = 1` in `iot_test_common_io_config.h`
3) Build, flash, and observe `aws_tests` SES project

Checklist:
----------


- [X] My code is Linted.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
